### ISSUE TITLE
Error in Chrome and Chromium based browsers: Manifest version 2 is deprecated

### ISSUE THREAD

**Issue by DaxMX** [frigus02/RESTer#874]
**The complete error messaje is:**

`Manifest version 2 is deprecated, and support will be removed in 2023. See https://developer.chrome.com/blog/mv2-transition/ for more details.`

So, this mean that the extension will stop working at any time?

**Comment by DaxMX** [comment]
Forgot to attach an screeshot.
![rester-manifest-v2-deprecated-error](https://user-images.githubusercontent.com/82778475/220244843-22396c32-3fdf-4567-a323-f76ebff23d9c.png)

Thanks in advance.

**Comment by frigus02** [comment]
Does the extension already not work? This looks like it may soon stop working.

I kinda knew about this and started a manifest v3 migration in https://github.com/frigus02/RESTer/tree/manifest-v3, but stopped prioritising it because (a) some features don't work with manifest v3 and (b) it wasn't supported by Firefox, yet.

This is a good reminder, though. I should look into this again and see if I at least do the migratin for Chrome.

**Comment by DaxMX** [comment]
The extension still work, the warning is just displayed in the Opera and Edge extensions manager, in firefox no warning and works so well.

Thank you.

**Comment by uoyniXnaH** [comment]
Hello @frigus02 ,
This issue seems to be stuck for a long time. Since Chrome stops support for manifest v2, it has been a crucial issue for Chrome users. I would like to work on this, though I have few experience on web extension spec.

Firstly let me list up TODOs here according to PR #882 :
- Remove default request header content-type
  - TODO: can we remove content-type if the body is not formData?
- TODO: test merging of browser cookie and custom cookie header
- TODO: with v3 Firefox seems to upgrade all requests to http because of the changed Content Security Policy
- TODO: background pages (or service workers) are not persistent anymore. This means the port in rester.js can and will disconnect after a while of inactivity. How do we deal with this? Ideally I'd want to reconnect only when needed. But at that time we may have missed a few events (settingsChanged or dataChanged), so the page is out of date. We need to bring it up to date again.

Today (2023-03-14) Chrome and Firefox differ in the APIs they support for Manifest v3:
- Chrome requires `background.serice_worker`. Firefox requires `background.scripts`. Luckily RESTer doesn't need any serice worker specific API. The same script runs fine with either API.
- Chrome dropped support for `webRequest` and requires `declarativeNetRequest`. Firefox does not support `declarativeNetRequest`, yet, but still supports `webRequest`.

For the second difference, I checked [MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/declarative_net_request) and it shows that Firefox started to support `declarativeNetRequest` since 2023/5/9. I guess it is no longer a problem.

**Comment by frigus02** [comment]
Yeah, this is tricky. Supporting manifest v3 is not trivial.

- Some parts have relatively clear solutions like the persistent background page. But this still feels like quite a bit of work to get right.
- Some parts are less clear. Like how to deal with features that don't work anymore.
- Some parts are just a bit ugly. Since Firefox continues to support `webRequest` I imagine it's going to support more features. It'd be a share to drop support because Chrome doesn't support it anymore. So this might mean supporting different features in different browers.

I'd like RESTer to keep working in new browser releases. But I honestly don't know when I'll have time to finish this. It also feels like might be a tricky piece to contribute. Although if you're willing to take a look and you manage to split this into reviewable chunks, maybe this can work. I'm not sure.

I realize that this isn't a very satisfying response. But it is what it is.

**Comment by uoyniXnaH** [comment]
Thank you for your response.
Since RESTer is still available on Chrome and Opera, I would like to start with some clear parts. Meanwhile we could list up those less clear ones.
But before that, I would like to know among those TODOs, what are features that not available in Manifest v3, and what are just for better performance or compatibility?

**Comment by frigus02** [comment]
I just looked at the PR again. I think it's [these](https://github.com/frigus02/RESTer/blob/93ad703f246d86d71716c20353b0386d3fd9638e/src/site/elements/data/scripts/request.js#L1-L7):

- Read the original resonse statusCode, statusText and headers.
  - I'm trying to remember what the issue was. I think with `webRequest` we can read the response pretty much as the server sent it, before the browser does any processing. But with `declarativeNetRequest` we can only specify some declarative rules on how to transform it. We have to figure out which headers we can't read like this. Might be stuff like `set-cookie` or maybe even `location` or `access-control-*`.
- Read original timing-allow-origin response header.
  - I think RESTer overwrites the header to make it so it can read request timing data. With `webRequest` it could move the original value into a different header. I imagine that's no longer possible with `declarativeNetRequest`.
- Append response header 'access-control-expose-headers' for all responses headers. Can't use '*' here because it's only a wildcard in requests without credentials.
- declarativeNetRequest doesn't seem to work with optional_permissions
  - This is probably just annoying but not a blocker.

**Comment by uoyniXnaH** [comment]
Sorry for my late reply.
I suppose that currently RESTer is using `webRequest`, thus migrating to `declarativeNetRequest` requires some modifications. I would like to study on `webRequest` and `declarativeNetRequest` and try solving issues you listed above.

**Comment by rhedin** [comment]
I guess it has turned the corner. 

I downloaded the repository, built it, and tried to load it using the "Load unpacked" button in Chrome.   Chrome refused. 

<img width="504" height="194" alt="Image" src="https://github.com/user-attachments/assets/dbb139f1-83c5-4a14-9ff9-b3a8752b1f6e" />

**Comment by frigus02** [comment]
Yeah, sorry. The migration turned out to be too much work. The new APIs are not as flexible and finding workarounds or removing features selectively in chrome was difficult.
